logic/tool/TypeResolver.java

```diff
@@ -373,88 +382,72 @@ private void convertValue(TypeVariable owner, ValueConstraint<?> constraint) {
+            for (TypeConstraint constraint : from.constraints()) {
+                if (constraint.isSub()) addNeighbours(from, constraint.asSub().type());
+                else if (constraint.isOwns()) addNeighbours(from, constraint.asOwns().attribute());
+                else if (constraint.isPlays()) addNeighbours(from, constraint.asPlays().role());
+                else if (constraint.isRelates()) addNeighbours(from, constraint.asRelates().role());
+                else if (constraint.isIs()) addNeighbours(from, constraint.asIs().variable());
+                // TODO: There are other constraints in a TypeVariable. Are we intentionally ignoring them?
+                //       Why do we not throw GraknException.of(ILLEGAL_STATE); ?
             }
         }
-
     }
 
-    private static class ResolverVariables {
+    private static class Resolvers {
 
-        private final Map<Identifier, TypeVariable> resolveeVars;
-        private final Map<RelationConstraint.RolePlayer, TypeVariable> rolePlayerResolveVars;
+        private final Map<Identifier, TypeVariable> typeResolvers;
+        private final Map<RelationConstraint.RolePlayer, TypeVariable> roleTypeResolvers;
+        // TODO: Why is there no checkers and getters for 'roleTypeResolvers'?
+        // TODO: do we still need 'roleTypeResolvers' given that we no longer put resolveRoleTypes on RelationConstraints?
 
-        private Integer tempCounter;
+        private int sysVarCounter;
 
-        ResolverVariables() {
-            this.tempCounter = 0;
-            this.resolveeVars = new HashMap<>();
-            this.rolePlayerResolveVars = new HashMap<>();
+        Resolvers() {
+            this.sysVarCounter = 0;
+            this.typeResolvers = new HashMap<>();
+            this.roleTypeResolvers = new HashMap<>();
         }
 
-        public TypeVariable convert(RelationConstraint.RolePlayer key) {
-            if (!rolePlayerResolveVars.containsKey(key)) {
-                TypeVariable newTypeVar = new TypeVariable(Identifier.Variable.of(
-                        new SystemReference("temp" + addAndGetCounter())));
-                rolePlayerResolveVars.put(key, newTypeVar);
-            }
-            return rolePlayerResolveVars.get(key);
+        TypeVariable register(RelationConstraint.RolePlayer rolePlayer) {
+            return roleTypeResolvers.computeIfAbsent(rolePlayer, rp -> new TypeVariable(newSystemVariable()));
         }
 
-        public TypeVariable convert(Variable key) {
-            if (!resolveeVars.containsKey(key.identifier())) {
+        TypeVariable register(Variable variable) {
+            return typeResolvers.computeIfAbsent(variable.identifier(), id -> {
                 TypeVariable newTypeVar;
-                if (key.reference().isAnonymous()) {
-                    newTypeVar = new TypeVariable(Identifier.Variable.of(
-                            new SystemReference("temp" + addAndGetCounter())));
```

Comment:
The fact that you need to prefix the argument of the `SystemReference` constructor with a `"temp"` string is a big code smell. Why do we need to this? I've pushed this logic down into the constructor method body (as it's always repeating), but this feels like something you should not be doing.

Reply:
We don't need to. I just did this for my own convenience to remind me how it was created.